matrix/plugins/glitch/selectors.py

```diff
@@ -0,0 +1,108 @@
+import functools
+import logging
+import random
+from typing import List, Any
+
+import enforce
+from juju.model import Model
+from juju.application import Application
+from juju.unit import Unit
+
+from matrix.utils import Singleton
+
+
+_marker = object()
+log = logging.getLogger("glitch")
+
+
+class _Selectors(dict, metaclass=Singleton):
+    def decorate(self, f):
+        wrapper = enforce.runtime_validation(f)
```

Comment:
I'm always wary when I see type checking in Python -- it feels like an argument for not using Python :-)

In this case, I had already begun to add some loose "guarantees" to the code, and this does so more formally, without needing to trust the programmer as much. I think that it makes sense, even if it does make me want to just port the project to Scala :-)